### ISSUE TITLE
Polymorphic serialization hangs

### ISSUE THREAD

**Issue by ukarlsson** [charleskorn/kaml#179]
Hello. Many thanks for this great library. We ran into an issue that is quite strange. The deserialization hangs in this specific case, meaning that the process does not terminate, but uses 100% CPU.

```
import com.charleskorn.kaml.Yaml
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer

@Serializable
private sealed class K {
    @Serializable
    @SerialName("x")
    data class X(
        val property: List<String>? = null,
    ) : K()
}

const val s = """
- !<x>
"""

fun main() {
    Yaml.default.decodeFromString(ListSerializer(K.serializer()), s)
}
```

This is what I see when I pause and enter debugger

![image](https://user-images.githubusercontent.com/2630658/131899310-2ef9d2a5-e04f-4f3a-87d6-430dbed4bf16.png)

**Comment by charleskorn** [comment]
Thanks for the bug report @ukarlsson. Are you able to share a sample project that triggers this issue? 

If not, could you please share what version of kaml, Kotlin and kotlinx.serialization you're using?

**Comment by ukarlsson** [comment]
Latest versions that I could find. I attached project

[kaml-bug.tar.gz](https://github.com/charleskorn/kaml/files/7109725/kaml-bug.tar.gz)

**Comment by ukarlsson** [comment]
I tried  to add a unit test, but there is some issue with gradle. I can neither run i IntelliJ nor in batect. Get this error

```
$ [main] ./batect check
Running check...
build-env: running ./gradlew check

To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/7.2/userguide/gradle_daemon.html#sec:disabling_the_daemon.
Daemon will be stopped at the end of the build 

> Configure project :
Kotlin Multiplatform Projects are an Alpha feature. See: https://kotlinlang.org/docs/reference/evolution/components-stability.html. To hide this message, add 'kotlin.mpp.stability.nowarn=true' to the Gradle properties.


FAILURE: Build failed with an exception.

* Where:
Build file '/code/build.gradle.kts' line: 28

* What went wrong:
An exception occurred applying plugin request [id: 'org.jetbrains.kotlin.jvm', version: '1.5.30']
> Failed to apply plugin 'org.jetbrains.kotlin.jvm'.
   > Cannot add extension with name 'kotlin', as there is an extension already registered with that name.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.2/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 4s
6 actionable tasks: 6 up-to-date

check finished with exit code 1 in 7.6s.
```

**Comment by charleskorn** [comment]
Thanks for the sample project.

I've been able to reproduce the issue, and slightly simplify the scenario that triggers it:

```kotlin
import com.charleskorn.kaml.Yaml
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
private sealed class K {
    @Serializable
    @SerialName("x")
    data class X(
        val property: String? = null,
    ) : K()
}

const val s = """
!<x>
"""

fun main() {
    println("Started.")
    val result = Yaml.default.decodeFromString(K.serializer(), s)
    println("Finished, result is $result")
}
```

The hang doesn't happen when `PolymorphismStyle.Property` is used - it only affects tagged polymorphism.

**Comment by charleskorn** [comment]
Thanks again for reporting this issue @ukarlsson. I've fixed it and [0.35.3](https://github.com/charleskorn/kaml/releases/tag/0.35.3) includes the fix.

Note that this may be a vulnerability depending on how your application uses kaml - please see [the security advisory](https://github.com/charleskorn/kaml/security/advisories/GHSA-fmm9-3gv8-58f4) for more information.

**Comment by ukarlsson** [comment]
Many thanks for fixing quickly!
